h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostUtils.java

```diff
@@ -112,20 +112,17 @@ else if (f.vec(i-catCols).isInt())
     public static long sumChunksLength(int[] chunkIds, Vec vec, Vec weightsVector, int[] chunkLengths) {
         for (int i = 0; i < chunkIds.length; i++) {
             final int chunk = chunkIds[i];
-            chunkLengths[i] = vec.chunkLen(chunk);
-            if (weightsVector == null)
-                continue;
-
-            Chunk weightVecChunk = weightsVector.chunkForChunkIdx(chunk);
-            if (weightVecChunk.atd(0) == 0) chunkLengths[i]--;
-            int nzIndex = 0;
-            do {
-                nzIndex = weightVecChunk.nextNZ(nzIndex, true);
-                if (nzIndex < 0 || nzIndex >= weightVecChunk._len) break;
-                if (weightVecChunk.atd(nzIndex) == 0) chunkLengths[i]--;
-            } while (true);
+            if (weightsVector != null) {
+                Chunk weightVecChunk = weightsVector.chunkForChunkIdx(chunk);
+                int nzIndex = 0;
+                do {
+                    if (weightVecChunk.atd(nzIndex) != 0) chunkLengths[i]++;
```

Comment:
For @michalkurka Here, only the zero elements should be subtracted. However, if you have a careful look, the logic here (introduced by me originally), it only subtracts the non-zero elements. Which means the final count is WRONG. (in the original code, honza's code does it vice versa and it works).